### PR TITLE
User convert

### PR DESCRIPTION
**Pull request by wrodriguez-dbx**
Remaining code for the `convert_user_to_service_account` usecase. Includes:
 - New service account service and repository.
 - New user repository.
 - New group edge repository.
 - New test cases.

And lots of other plumbing.

Notes/questions:

 - Starting to run into a bit of an issue of "repos use lots of other repos", which could later lead to circular dependencies. Not sure how to better segment things to avoid the issue, or if it's fine as is.

 - entities/ seems like a good place to plop exception definitions. I think my only exception (heh) to that is in the User repo, but can move those too if you agree.

 - In some places there is code structure that is now unnecessary because I changed approaches part way through; see for example the `add_user_to_group` functions. I figured might as well leave those in since guessing they'll be necessary later, but arguably should be removed for cleanliness.

 - Currently there's nothing testing edge case checking at levels below the usecase, so in particular nothing exercises the edge case checking code at the repo layer. Should that change?

### CODE REVIEW

File: grouper/entities/service_account.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+class ServiceAccountNotFoundException(Exception):
+    """Attempt to operate on a service account not found in the storage layer."""
+
+    def __init__(self, name):
+        # type: (str) -> None
+        msg = "Service account {} not found".format(name)
+        super(ServiceAccountNotFoundException, self).__init__(msg)
+
+
+class ServiceAccountHasOwnerException(Exception):
```
Comment (rra) [inline grouper/entities/service_account.py]: I don't think we'll ever need this exception, since it's not a constraint violation at the database layer to change the owner of a service account. The use case may want to check various things before taking this action, but I don't think the service or repository should reject it.

File: grouper/entities/user.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
+from typing import NamedTuple
+
+User = NamedTuple("User", [("id", int), ("username", str), ("enabled", bool)])
```
Comment (rra) [inline grouper/entities/user.py]: We should never need id and shouldn't expose it below the repository layer. In general, we should only expose id if there's no other way of talking about a specific instance of an entity. Users and groups have unique names, so the rest of the code should just use the name.

I would prefer `name` to `username`. `user.username` is sort of irritatingly redundant and doesn't add any clarity.

File: grouper/repositories/audit_log.py [hunk 3]
```diff
@@ -16,12 +19,6 @@
     from typing import List, Optional
 
 
-class UnknownGroupException(Exception):
-    """Group involved in a logged action does not exist."""
-
-    pass
-
-
 class UnknownPermissionException(Exception):
```
Comment (rra) [inline grouper/repositories/audit_log.py]: I think you wanted to get rid of this one too, since you're importing another version of it above.

File: grouper/repositories/group_edge.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+if TYPE_CHECKING:
+    from grouper.graph import GroupGraph
+    from grouper.models.base.session import Session
+    from typing import List
+
+
+class GraphGroupEdgeRepository(GroupEdgeRepository):
+    """Graph-aware storage layer for group edges.group
+
+    WARNING: This repository should only be used from other repositories. At higher levels -- the
```
Comment (rra) [inline grouper/repositories/group_edge.py]: Looking at how this code turned out, I think we should avoid having this additional layer. The service layer should make the calls to various repositories, and repositories should never call each other. So this abstraction would be provided by the service layer, rather than at the repository layer.

File: grouper/repositories/group_edge.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        # type: (str, str, str) -> None
+        raise NotImplementedError()
+
+    def groups_of_user(self, username):
+        # type: (str) -> List[str]
+        user_details = self.graph.get_user_details(username)
+        return [group for group in user_details["groups"]]
+
+
+class SQLGroupEdgeRepository(GroupEdgeRepository):
```
Comment (rra) [inline grouper/repositories/group_edge.py]: I have mixed feelings about leaving in this much unused code. I suppose it's not hurting anything, but I would lean towards only writing the code required for the use case and leaving the rest to be filled in later.

File: grouper/repositories/group_request.py [hunk 6]
```diff
@@ -40,39 +40,25 @@ def __init__(self, session):
...
-        self.session.flush()
-
-        Comment(
-            obj_type=OBJ_TYPES["RequestStatusChange"],
-            obj_pk=request_status_change.id,
-            user_id=actor.id,
-            comment=reason,
-            created_on=now,
-        ).add(self.session)
+        request.update_status(actor, "cancelled", reason)
```
Comment (rra) [inline grouper/repositories/group_request.py]: Please undo this change -- it's intentional that the repository layer does not use any of these methods on the model. The model methods are going away and shouldn't be used by any new code.

File: grouper/repositories/group_request.py [hunk 7]
```diff
@@ -40,39 +40,25 @@ def __init__(self, session):
...
 
     def pending_requests_for_user(self, user):
         # type: (str) -> List[UserGroupRequest]
         requester = aliased(User)
         on_behalf_of = aliased(User)
         sql_requests = self.session.query(
             Request.id,
+            Request.status,
```
Comment (rra) [inline grouper/repositories/group_request.py:64]: You shouldn't need this.

File: grouper/repositories/interfaces.py [hunk 8]
```diff
@@ -9,6 +9,26 @@
     from typing import List, Optional
 
 
+class GroupEdgeRepository(object):
+    """Abstract base class for group edge repositories.class
```
Comment (rra) [inline grouper/repositories/interfaces.py]: Stray word at the end here. Also drop the warning comment per above.

File: grouper/repositories/interfaces.py [hunk 9]
```diff
@@ -44,3 +64,64 @@ def permissions_for_user(self, user):
     def user_has_permission(self, user, permission):
         # type: (str, str) -> bool
         pass
+
+
+class ServiceAccountRepository(object):
```
Comment (rra) [inline grouper/repositories/interfaces.py]: I'm not sure it's worth creating interfaces and both graph and non-graph versions of these repositories now when we don't need them yet. I was avoiding the interface layer for new repositories until I really needed it, since in the long run we may be getting rid of, or at least greatly reducing, the scope of the graph.

File: grouper/repositories/interfaces.py [hunk 10]
```diff
@@ -44,3 +64,64 @@ def permissions_for_user(self, user):
...
+    def disable_user(self, username):
+        # type: (str) -> None
+        pass
+
+    @abstractmethod
+    def enable_user(self, username):
+        # type: (str) -> None
+        pass
+
+    @abstractmethod
```
Comment (rra) [inline grouper/repositories/interfaces.py]: Per above, I'd get rid of this method on the user repository entirely and have the service call the group edge repository for this information.

File: grouper/repositories/service_account.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+        # type: (str, str) -> None
+        service_account = SQLServiceAccount.get(self.session, name=name)
+        if not service_account:
+            raise ServiceAccountNotFoundException(name)
+
+        group = Group.get(self.session, name=groupname)
+        if not group:
+            raise GroupNotFoundException(groupname)
+
+        existing_relationship = GroupServiceAccount.get(
```
Comment (rra) [inline grouper/repositories/service_account.py:33]: I'd drop this check and let this function change the ownership. Repositories (and services) should do anything allowed by the schema; checking whether that action is correct is up to the use case.

File: grouper/repositories/interfaces.py [hunk 12]
```diff
@@ -44,3 +64,64 @@ def permissions_for_user(self, user):
...
+    def groups_of_user(self, username):
+        # type: (str) -> List[str]
+        pass
+
+    @abstractmethod
+    def mark_disabled_user_as_service_account(self, username):
```
Comment (rra) [inline grouper/repositories/interfaces.py]: This should be a function on the service account repository. (And the comment should be a docstring.)

File: grouper/repositories/service_account.py [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+
+    def enable_service_account(self, name):
+        # type: (str) -> None
+        # NOTE: The fact that we have this check that the service account exists and then
+        # never use it is an artifact of service accounts being built on top of users.
+        # It can be fixed once the two repositories are properly separate.
+        service_account = SQLServiceAccount.get(self.session, name=name)
+        if not service_account:
+            raise ServiceAccountNotFoundException(name)
+        return self.user_repository.enable_user(name)
```
Comment (rra) [inline grouper/repositories/service_account.py]: I don't think the service account repository should be aware of, or call into, the user repository. Instead, I think this should say:

```
service_account.user.enabled = True
```

This keeps all of the code localized to the service account repository and means nothing about the structure of the classes has to change when we refactor the database schema.

File: grouper/repositories/user.py [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+
+    def enable_user(self, username):
+        # type: (str) -> None
+        user = SQLUser.get(self.session, name=username)
+        if not user:
+            raise UserNotFoundException(username)
+
+        user.enabled = True
+
+    def groups_of_user(self, username):
```
Comment (rra) [inline grouper/repositories/user.py]: As above, I'd get rid of this method.

File: grouper/repositories/user.py [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+        # type: (str) -> List[str]
+        return self.group_edge_repository.groups_of_user(username)
+
+    def mark_disabled_user_as_service_account(self, username):
```
Comment (rra) [inline grouper/repositories/user.py]: As above, I think this should be a method on `ServiceAccountRepository`.

File: grouper/repositories/user.py [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+        # type: (str) -> None
+        user = SQLUser.get(self.session, name=username)
+        if not user:
+            raise UserNotFoundException(username)
+
+        if self.groups_of_user(username) != []:
```
Comment (rra) [inline grouper/repositories/user.py]: Seeing this code, I think this check and the one below should be done at the service layer, since they plug together multiple repositories. The repository should just do what it's told and not make this check. That mildly increases the chance of an inconsistency by some other service incorrectly calling into the repository, but I think the chances of that are slight since everything else in Grouper should be using the service.

File: grouper/services/service_account.py [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+
+    def create_service_account_from_disabled_user(self, user, authorization):
+        # type: (str, Authorization) -> None
+
+        # WARNING: This logic relies on the fact that the user and service account repos
+        # are in fact the same thing, as it never explicitly removes the user from the
+        # user repo. This is a temporary breaking of the abstractions and will have to be
+        # cleaned up once the repositories are properly separate.
+        self.user_repository.mark_disabled_user_as_service_account(user)
+        self.service_account_repository.set_service_account_description(user, "")
```
Comment (rra) [inline grouper/services/service_account.py]: Why these calls as separate calls? I think this has to be done by `mark_disabled_user_as_service_account` since otherwise that call isn't stand-alone.

File: grouper/services/service_account.py [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+        self.service_account_repository.set_service_account_mdbset(user, "")
+
+        self.audit_log.log_create_service_account_from_disabled_user(user, authorization)
+
+    def enable_service_account(self, user, owner, authorization):
+        # type: (str, str, Authorization) -> None
+        self.service_account_repository.assign_service_account_to_group(user, owner)
```
Comment (rra) [inline grouper/services/service_account.py:56]: This needs audit logging separate from enabling the service account.

File: grouper/services/service_account.py [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+        group_request_repository,  # type: GroupRequestRepository
+        audit_log_service,  # type: AuditLogService
+    ):
+        # type: (...) -> None
+        self.user_repository = user_repository
+        self.service_account_repository = service_account_repository
+        self.group_request_repository = group_request_repository
+        self.audit_log = audit_log_service
+
+    def create_service_account_from_disabled_user(self, user, authorization):
```
Comment (rra) [inline grouper/services/service_account.py:37]: Per above, I'd move the sanity checking down to here.

File: tests/setup.py [hunk 20]
```diff
@@ -147,3 +147,15 @@ def grant_permission_to_group(self, permission, argument, group):
             permission_id=permission_obj.id, group_id=group_obj.id, argument=argument
         )
         grant.add(self.session)
+
+    def create_group_request(self, user, group, role="member"):
```
Comment (rra) [inline tests/setup.py:151]: This doesn't do what the function name says so far as I can tell. It just adds the user directly to the group. (Maybe it creates a request as a side effect, but I think you instead want to create a request explicitly and directly.)

File: grouper/repositories/audit_log.py (hunk 3 above)
Comment (wrodriguez-dbx) [reply grouper/repositories/audit_log.py]: Yup, thanks.

File: grouper/repositories/group_edge.py (hunk 5 above)
Comment (wrodriguez-dbx) [reply grouper/repositories/group_edge.py]: Yea, it was a case of I wrote it, realized I didn't need it anymore, and didn't know whether to do delete it. Going back and deleting now to make this PR less of a beast.

File: grouper/repositories/user.py (hunk 16 above)
Comment (wrodriguez-dbx) [reply grouper/repositories/user.py]: Hmm, could we maybe say that anything that could cause an inconsistency gets checked at the service layer?

While it does feel a little weird (it's a database inconsistency, so why not check at the repo?), I think it ultimately makes sense. The consistency invariants would be defined by the service layer, which guarantees that those invariants are the same across all database backends.

Changed to use the above strategy; let me know how you feel about it.

File: grouper/services/service_account.py (hunk 17 above)
Comment (wrodriguez-dbx) [reply grouper/services/service_account.py]: This was me trying to be disciplined about what logic went to which repo: marking a disabled user a service account acts on a user, whereas setting the attributes of that service account is acting on a service account, so they went into separate repos (and thus separate calls).

In retrospect though, better to contain the messiness inside of the repo by being slightly less disciplined about separation. Changing.

File: grouper/services/service_account.py (hunk 18 above)
Comment (wrodriguez-dbx) [reply grouper/services/service_account.py:56]: Hmm, if we consider enabling a service account to include assigning its owner, is one audit log entry not more appropriate?  Feels like either one audit log entry is correct or this should be two separate functions at the service layer, so in either case one service layer function logs once.

File: tests/setup.py (hunk 20 above)
Comment (wrodriguez-dbx) [reply tests/setup.py:151]: This actually does only create the request; there's a default argument `status="pending"` in there. I'll make that explicit and add a comment. The flow that requests to join a group is so wonky that I wanted to preserve exact behavior by calling the function directly.

File: grouper/repositories/group_request.py (hunk 7 above)
Comment (wrodriguez-dbx) [reply grouper/repositories/group_request.py:64]: Without it the `status=sql_request.status` in creating the `UserGroupRequest` below fails (the result object doesn't have a status field).

File: grouper/repositories/interfaces.py (hunk 8 above)
Comment (wrodriguez-dbx) [reply grouper/repositories/interfaces.py]: Oops thanks.

File: grouper/repositories/interfaces.py (hunk 9 above)
Comment (wrodriguez-dbx) [reply grouper/repositories/interfaces.py]: Sure, makes sense. Trimming down as part of killing the unused code I wrote.

File: grouper/repositories/interfaces.py (hunk 12 above)
Comment (wrodriguez-dbx) [reply grouper/repositories/interfaces.py]: Sure. Logic for putting it in user repo was that it was acting on a user, but I don't have a strong opinion.

File: grouper/repositories/service_account.py (hunk 11 above)
Comment (wrodriguez-dbx) [reply grouper/repositories/service_account.py:33]: Sure, makes sense.

File: grouper/repositories/service_account.py (hunk 13 above)
Comment (wrodriguez-dbx) [reply grouper/repositories/service_account.py]: Oh I didn't know that `service_account.user` worked like that. That's helpful.

File: grouper/repositories/group_request.py (hunk 6 above)
Comment (wrodriguez-dbx) [reply grouper/repositories/group_request.py]: Ah, didn't know the immediate plan was to kill the model methods and there were minor issues here so I fixed them with a sledgehammer. Changing back.

File: grouper/repositories/group_edge.py (hunk 4 above)
Comment (wrodriguez-dbx) [reply grouper/repositories/group_edge.py]: Yea, I agree. Changed.

File: grouper/repositories/group_request.py (hunk 7 above)
Comment (rra) [reply grouper/repositories/group_request.py:64]: Oh, I see, thank you!

File: grouper/services/service_account.py (hunk 18 above)
Comment (rra) [reply grouper/services/service_account.py:56]: Ah, I see, this is including the group in the audit log message as well. Yeah, this seems fine to me.

File: tests/setup.py (hunk 20 above)
Comment (rra) [reply tests/setup.py:151]: Ah, I see. Yeah, this looks good.

**Review by rra — APPROVED** [review]
This all looks good to me. Thanks! I'll merge this into my branch and then rebase and squash and put up a final PR for dropbox/merou for you to look over.
